KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,742 @@
+            require(approvers[i] != _approver, "Already approved");
+        }
+        retirees[index].approvers.push(_approver);
+        emit Approved(
+            _retiredAddress,
+            _approver,
+            retirees[index].approvers.length
+        );
+    }
+
+    /**
+     * @dev finalizeRegistration sets the status to Registered,
+     *      After this stage, registrations will be restricted.
+     */
+    function finalizeRegistration()
+        public
+        onlyOwner
+        onlyAtStatus(Status.Initialized)
+    {
+        status = Status.Registered;
+        emit StatusChanged(status);
+    }
+
+    /**
+     * @dev finalizeApproval sets the status to Approved,
+     *      After this stage, approvals will be restricted.
+     */
+    function finalizeApproval()
+        public
+        onlyOwner
+        onlyAtStatus(Status.Registered)
+    {
+        require(
+            getTreasuryAmount() < sumOfRetiredBalance(),
+            "treasury amount should be less than the sum of all retired address balances"
+        );
+        _isRetiredsApproved();
+        status = Status.Approved;
+        emit StatusChanged(status);
+    }
+
+    /**
+     * @dev verify if quorom reached for the retired approvals
+     */
+    function _isRetiredsApproved() private view {
+        for (uint256 i = 0; i < retirees.length; i++) {
+            Retired memory retired = retirees[i];
+            bool isContract = isContractAddr(retired.retired);
+            if (isContract) {
+                (address[] memory adminList, uint256 req) = _getState(retired.retired);
+                if (retired.approvers.length >= req) {
+                    //if min quorom reached, make sure all approvers are still valid
+                    address[] memory approvers = retired.approvers;
+                    uint256 minApprovals = 0;
+                    for (uint256 j = 0; j < approvers.length; j++) {
+                        _validateAdmin(approvers[j], adminList);
+                        minApprovals++;
+                    }
+                    require(
+                        minApprovals >= req,
+                        "min required admins should approve"
+                    );
+                } else {
+                    revert("min required admins should approve");
+                }
+            }else {
+                require(retired.approvers.length == 1 , "EOA should approve");
+            }
+        }
+    }
+
+    /**
+     * @dev sets the status of the contract to Finalize. Once finalized the storage data
+     * of the contract cannot be modified
+     * @param _memo is the result of the rebalance after executing successfully in the core.
+     * @return retirees is an array of retired address
+     * @return totalRetireesBalance is the sum of all retired balances
+     * @return newbies is an array of newbie address
+     * @return totalNewbiesFund is the sum of funds allocated to
+     */
+    function finalizeContract(
```

Comment:
This function should be called by core.
@aidan-kwon Is there any mechanism that only core can call this function?

Reply:
There is a way, but we try to avoid "write" function call from the core side. let me test and see